### ISSUE TITLE
HistogramLUTItem sigLevelsChanged problem

### ISSUE THREAD

**Issue by Fokatu** [pyqtgraph/pyqtgraph#1512]
When I try to change the levels of an image of size 2048x2048 with HistogramLUTItem, there is obvious lag. Is this caused by too many “sigLevelsChanged” signals? Can we adjust the time interval between adjacent two signals?

**Comment by j9ac9k** [comment]
We have SignalProxy can do rate limiting:

https://github.com/pyqtgraph/pyqtgraph/blob/23a46b5fb98b4d0cfadbb2cdb74a08c6b06f667d/pyqtgraph/SignalProxy.py#L12-L20

I don't use HistogramLUTItem much so I have no opinion if this should be implemented or not; if you can provide a minimal example (this is usually easiest done by modifying one of the examples) that would be helpful.

**Comment by Fokatu** [comment]
@j9ac9k ，Thanks for your reply. The SignalProxy class works for me. But for bigger image, the problem appears again.

Here is a small example
`import numpy as np
from pyqtgraph.Qt import QtGui
import pyqtgraph as pg
pg.setConfigOptions(imageAxisOrder='row-major')

app = QtGui.QApplication([])

win = QtGui.QMainWindow()
win.resize(800,800)
imv = pg.ImageView()
win.setCentralWidget(imv)
win.show()

data = np.random.normal(size=(4096, 4096))
imv.setImage(data)

QtGui.QApplication.instance().exec_()
`

**Comment by j9ac9k** [comment]
For this sort of thing in my application I've put the emitting of a signals into a queue of just _one_ entry and use a `QThread` to grab what's in the queue and do the computation there.  I'm not sure if that's an appropriate implementation for the library; but it might be.

**Comment by Fokatu** [comment]
I have tried to update image in another thread. There will be no lag in the histogram widget, but it feel quite strange if the image update is not synchronous.

I have an idea, but I am not sure if it should be added into the library. You know, the generated qimage have only 256 gray scales, so I think for most of the level change signal, there should be no actual change in the qimage. Maybe we can add a extra if to deside whether sending the singal or not

**Comment by j9ac9k** [comment]
@outofculture @ixjlyons  do you all have thoughts on this?  I don't work w/ the LUTs much so I'm not sure what a solution here would look like.

**Comment by outofculture** [comment]
I don't think this is a problem of too many `sigLevelsChanged` signals. What evidence do you have of that being at fault, @Fokatu? My investigation didn't find that to be case, and in fact I could comment out the signal emit and things behaved the same.

In any case, this needs more investigation before it can be solved. Peppering more code with `pyqtgraph.debug.Profiler` statements might reveal something. At a glance, maybe the histogram needs to no longer directly call `setLevels` on its imageItem, or if it does, it should do so outside of the main UI thread?

Also, this issue should be renamed to no longer cast aspersions against a poor, innocent signal. :-P

**Comment by pijyoi** [comment]
If you comment out the following line, it seems to solve the lag:
```python
region.sigRegionChanged.connect(self.regionChanging)
```
https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/HistogramLUTItem.py#L89

**Comment by outofculture** [comment]
Ah, I see, thank you.

Removing that line also changes the behavior to no longer live-update the image. That change in behavior could maybe be a setting? We should still investigate solutions that don't change things quite so much.

**Comment by pijyoi** [comment]
Instead of disabling the signal outright, the same effect could be achieved by changing the following line:
https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/HistogramLUTItem.py#L210
to:
```python
self.imageItem().setLevels(self.getLevels(), update=False)
```

So the issue is not so much with the signal per se, but with the long ImageItem::render() time needed with large images. (On my system, render() was taking 750ms.) Because render() runs in the GUI thread, the user perceives lag in their histogram mouse interactions.

As you suggested, this update keyword could be a setting for users manipulating large images (and/or running slower systems)

Optimizing render() would also help.

**Comment by j9ac9k** [comment]
I think @dgoeries talked about some optimization that could be had in applyLevels fairly recently.

**Comment by pijyoi** [comment]
Some optimizations are suggested in #1518

**Comment by pijyoi** [comment]
@Fokatu, np.random.normal() as used in your example is only able to output float64. Are your actual data types float64 or are you able to use lower precision types like float32?

**Comment by Fokatu** [comment]
Hi, @pijyoi ，I may meet data with different types.  By using both the SignalProxy class and the trick that sending signals only when  the levels change is large enough, the performance is acceptable now.
But still, if compared with imageJ, PQ is slow. Can we render the image with multi-processing library?

**Comment by pijyoi** [comment]
@Fokatu , you could turn on auto downsampling for large images.
```python
imv.getImageItem().setAutoDownsample(True)
imv.setImage(data)
```

**Comment by outofculture** [comment]
If you have the hardware, you could also send your data to the GPU and use CUDA processing:

```python
import cupy as cp
pg.setConfigOptions(useCupy=True)

# ...

imv.setImage(cp.asarray(data))
# or generate data using cupy to begin with
```

**Comment by j9ac9k** [comment]
I can't help but wonder if there is a faster implementation to be had here:

https://github.com/pyqtgraph/pyqtgraph/blob/8670407779ede5f6e1c0305266508608945d286e/pyqtgraph/functions.py#L1095-L1125

**Comment by pijyoi** [comment]
It seems that the cumulative optimizations up to and including #1648 makes the example here using the (4096, 4096) float64 image run tolerably (even on Windows).

Although the source data set is float64, the target display array is ubyte, thus the fastpath is taken.

**Comment by pijyoi** [comment]
@Fokatu , you could try the following with current master:
```python
import pyqtgraph as pg
import numpy as np
pg.setConfigOptions(imageAxisOrder='row-major')
pg.setConfigOptions(useNumba=True)

app = pg.mkQApp()
imv = pg.ImageView()
imv.show()
size = (4096, 4096)
data = np.random.normal(size=size).astype(np.float32)
imv.setImage(data)
app.exec_()
```
